### PR TITLE
Decoupling

### PR DESCRIPTION
**Pull request by ZmagoD**

### CODE REVIEW

File: app/views/protocols/index.html.erb [hunk 1]
```diff
@@ -49,14 +49,28 @@
...
-            data-team-id="<%= @current_team.id %>"
-            data-type="<%= @type %>" data-import-url="<%= import_protocols_path %>"
-            <%= 'disabled="disabled"' unless can_create_protocols_in_repository?(@current_team) %>>
-          </a>
+        <a class="btn btn-default btn-open-file" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" href="#">
+          <span class="glyphicon glyphicon-import"></span><span class="hidden-xs">&nbsp;<%= "Import protocol" %></span></a>
+
+          <ul class="dropdown-menu">
+            <li>
+              <a class="btn btn-default btn-open-file" <%= can_create_protocols_in_repository?(@current_team) ? 'data-action="import"' : 'disabled="disabled"' %>>
```
Comment (mlorb) [inline app/views/protocols/index.html.erb]: Remove `btn btn-default` class

File: app/helpers/permission_helper.rb [hunk 2]
```diff
@@ -844,42 +843,42 @@ def can_link_copied_protocol_in_repository(protocol)
...
+    elsif protocol.in_repository? # DONE
       protocol.in_repository_active? and can_edit_protocol(protocol)
     else
       false
     end
   end
 
-  def can_reorder_step_in_protocol(protocol)
-    if protocol.in_module?
+  def can_reorder_step_in_protocol(protocol) # WIP
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/helpers/permission_helper.rb]: we will clean this when we refactor permissions

File: app/helpers/permission_helper.rb [hunk 3]
```diff
@@ -892,28 +891,28 @@ def can_reorder_step_in_protocol(protocol)
   #   - adding assets
   #   - adding tables
   # but right now we have 1 page to rule them all.
-  def can_edit_step_in_protocol(protocol)
-    if protocol.in_module?
+  def can_edit_step_in_protocol(protocol) # WIP
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/helpers/permission_helper.rb]: we will clean this when we refactor permissions

File: app/helpers/permission_helper.rb [hunk 4]
```diff
@@ -892,28 +891,28 @@ def can_reorder_step_in_protocol(protocol)
...
+    elsif protocol.in_repository? # DONE
       protocol.in_repository_active? and can_edit_protocol(protocol)
     else
       false
     end
   end
 
-  def can_delete_step_in_protocol(protocol)
-    if protocol.in_module?
+  def can_delete_step_in_protocol(protocol) # WIP
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/helpers/permission_helper.rb]: we will clean this when we refactor permissions

File: app/helpers/permission_helper.rb [hunk 5]
```diff
@@ -720,45 +719,45 @@ def can_view_protocol(protocol)
...
+  # end
+
+  # def can_publish_protocol(protocol)
+  #   protocol.added_by == current_user and
+  #   protocol.in_repository_private?
+  # end
+
+  def can_export_protocol(protocol) # WIP
+    (protocol.in_repository_public? and is_member_of_team(protocol.team)) or # DONE
+      (protocol.in_repository_private? and protocol.added_by == current_user) or # DONE
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/helpers/permission_helper.rb]: we will clean this when we refactor permissions

File: app/helpers/permission_helper.rb [hunk 6]
```diff
@@ -720,45 +719,45 @@ def can_view_protocol(protocol)
...
+  #   protocol.in_repository_public?
+  # end
+
+  # def can_publish_protocol(protocol)
+  #   protocol.added_by == current_user and
+  #   protocol.in_repository_private?
+  # end
+
+  def can_export_protocol(protocol) # WIP
+    (protocol.in_repository_public? and is_member_of_team(protocol.team)) or # DONE
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/helpers/permission_helper.rb]: we will clean this when we refactor permissions

File: app/helpers/permission_helper.rb [hunk 7]
```diff
@@ -720,45 +719,45 @@ def can_view_protocol(protocol)
...
-    protocol.in_repository_private?
-  end
-
-  def can_export_protocol(protocol)
-    (protocol.in_repository_public? and is_member_of_team(protocol.team)) or
-      (protocol.in_repository_private? and protocol.added_by == current_user) or
-      (protocol.in_module? and
+  # def can_edit_protocol(protocol)
+  #   is_normal_user_or_admin_of_team(protocol.team) and
+  #   current_user == protocol.added_by and (not protocol.in_repository_archived?)
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/helpers/permission_helper.rb]: we will clean this when we refactor permissions

File: app/models/concerns/user/project_roles.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+  def is_creator_of_project?(project)
+    project.created_by == self
+  end
+
+  def is_owner_of_project?(project)
+    @user_project.owner?
+  end
+
+  def is_user_of_project?(project)
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/models/concerns/user/project_roles.rb]: :banana: we will clean this when we refactor permissions

File: app/models/concerns/user/project_roles.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    @user_project.normal_user?
+  end
+
+  def is_user_or_higher_of_project?(project)
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/models/concerns/user/project_roles.rb]: :banana: we will clean this when we refactor permissions

File: app/services/client_api/user_team_service.rb [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                            user,
+                            @user_team.team,
+                            false,
+                            false)
+    end
+
+    def validate_params(args)
+      keys = %i(team_id user_team_id user)
+      raise ClientApi::CustomUserTeamError unless keys.all? { |s| args.key? s }
+      raise ClientApi::CustomUserTeamError if args.values.any? &:nil?
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/services/client_api/user_team_service.rb]: Do we still care about this?

File: app/services/client_api/invitations_service.rb [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    def handle_user(result, email, user)
+      return handle_new_user(result, email, user) if user.blank?
+      handle_existing_user(result, user)
+    end
+
+    def handle_new_user(result, email, user)
+      password = generate_user_password
+      # Validate the user data
+      error = (Constants::BASIC_EMAIL_REGEX !~ email)
+      error = validate_user(email, email, password).count > 0 unless error
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/services/client_api/invitations_service.rb]: Do we still care about this?

File: app/permissions/team.rb [hunk 12]
```diff
@@ -0,0 +1,119 @@
+Canaid::Permissions.register_for(Team) do
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/permissions/team.rb]: Do we still care about this?

**Pull request merged**